dask/array/tests/test_routines.py

```diff
@@ -1286,3 +1286,37 @@ def test_result_type():
     # dask 0d-arrays are NOT treated like scalars
     c = da.from_array(np.ones((), np.float64), chunks=())
     assert da.result_type(a, c) == np.float64
+
+
+@pytest.mark.parametrize('einsum_signature', [
+    'abc,bad->abcd',
+    'abcdef,bcdfg->abcdeg',
+    'ea,fb,abcd,gc,hd->efgh',
```

Comment:
it would be good to add a few more examples for more test coverage:
- with outputs omitted, e.g., `'ab,b'`
- with no output indices, e.g., `'ab,b->'
- `...` for broadcasting, e.g., `'...a,a'` (could raise an informative error for now)
- with the alternative syntax `einsum(op0, sublist0, op1, sublist1, ..., [sublistout])` (could raise an informative error for now)

Reply:
As a note I would also explicitly test broadcasting operations such as 
```
a.shape = (4, 1)
b.shape = (4, 4)
np.einsum('ij,jk->ik', a, b)
```

A few more similar tests can be found [here](https://github.com/dgasmith/opt_einsum/blob/master/opt_einsum/tests/test_contract.py#L247).

Einsums functionality is incredibly broad and I got caught on a few issues that were not tested in NumPy and as such NumPy 14.0 had a few regressions from my changes. A few other tricky cases I would recommend:

```
'aa->a`,
'ab,ab,c->c'   
'aab,bc->ac',
'aab,bcc->ac',
'fdf,cdd,ccd,afe->ae',
'fff,fae,bef,def->abd',
```

Reply:
@dgasmith OK, I've added the test cases in your comment as well as the broadcasting test cases. They succeed locally for me and I've pushed them up.

I realise that the range of einsum's behaviour is broad, but my available time for working on this PR is running out.

With regard to the dask wrapper,  as long as the full subscript specificaiton is present, dask will align each chunk correctly and defer to `np.einsum`'s implementation